plugins/cilium-cni/cmd/cmd.go

```diff
@@ -210,15 +210,42 @@ func allocateIPsWithDelegatedPlugin(
 	// Safe to assume at most one IP per family. The K8s API docs say:
 	// "Pods may be allocated at most 1 value for each of IPv4 and IPv6"
 	// https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/
+	// Interface returned by IPAM should be treated as the uplink for the Pod as CNI spec introduced by:
+	// https://github.com/containernetworking/cni/pull/1137
+	masterMac := ""
+	for _, iface := range ipamResult.Interfaces {
+		if iface.Sandbox == "" && iface.Name != "" {
+			if uplink, err := netlink.LinkByName(iface.Name); err != nil {
+				log.WithError(err).WithField("uplink", iface.Name).Error("Failed to get uplink")
```

Comment:
I have two points as feedback here:

1. This needs to be `safenetlink.LinkByName()` (see its godoc for details)
2. In the past (and probably even present), we had problems on ENI mode where the cloud provider (i.e. the Amazon EC2 API in our case) would attach a new interface and an IP to the local node (and thus became available in IPAM), but the interface itself was yet discovered by the local Linux kernel. Thus causing such `LinkByName` lookups to fail until the interface was actually attached.

I think it would be worth thinking about how this code deals with such cases (IPAM plugin belives an interface exist, but the local kernel has not seen it yet, because it's still in the process of being attached): Do we return an error from CNI ADD and let kublet retry later (hopefully at a time when the interface will exist), or do we attempt to wait here?

The first one (letting CNI ADD fail) is probably the easier to implement. For that to work, you will have to return an error here and make sure CNI ADD fails, rather than just logging it here.

Reply:
Thanks @gandro, it should return error if link could not be found yet.